src/Aspire.Dashboard/Model/IDashboardClientExtensions.cs

```diff
@@ -0,0 +1,36 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Globalization;
+
+namespace Aspire.Dashboard.Model;
+
+internal static class IDashboardClientExtensions
+{
+    /// <summary>
+    /// Gets the application name if available. If it's not yet available, an empty string is returned
+    /// synchronously, and <paramref name="refresh"/> is invoked later on, when the value is available.
+    /// The expectation is that callers pass <c>() => InvokeAsync(StateHasChanged)</c>.
+    /// </summary>
+    public static string FormatApplicationName(this IDashboardClient client, string pattern, Action refresh)
+    {
+        var task = client.WhenConnected;
+
+        if (!task.IsCompleted)
+        {
+            // We're not yet connected. Schedule refresh to be invoked
+            // when the connection is established.
+            task.ContinueWith(
```

Comment:
Do we care whether refresh() is called only once after task completion? I wonder what happens if FormatApplicationName() is called multiple times...

Reply:
`FormatApplicationName` is called twice for each page -- once for the `<title>` element and once for the heading shown in the UI. If both require the continuation to trigger refresh, that refresh would happen twice, though in both cases on different parts of the UI, as I understand it. So I don't foresee any issue here.

Reply:
I'm going to explore James's suggestion of making a component for this.

Reply:
https://github.com/dotnet/aspire/pull/1555